solver/src/naive_solver/multi_order_solver.rs

```diff
@@ -22,18 +22,55 @@ impl TokenContext {
                 * u256_to_bigint(&deficit.reserve)
                 * (u256_to_bigint(&self.sell_volume) - u256_to_bigint(&self.buy_volume))
     }
+
+    pub fn is_excess_before_fees(&self, deficit: &TokenContext) -> bool {
+        u256_to_bigint(&self.reserve)
+            * (u256_to_bigint(&deficit.sell_volume) - u256_to_bigint(&deficit.buy_volume))
+            < u256_to_bigint(&deficit.reserve)
+                * (u256_to_bigint(&self.sell_volume) - u256_to_bigint(&self.buy_volume))
+    }
+}
+
+pub fn solve(
+    orders: impl Iterator<Item = OrderCreation> + Clone,
+    pool: &Pool,
+) -> SinglePairSettlement {
+    let mut orders: Vec<OrderCreation> = orders.collect();
+    while !orders.is_empty() {
+        let (context_a, context_b) = split_into_contexts(orders.clone().into_iter(), pool);
+        let solution = solve_orders(orders.clone().into_iter(), &context_a, &context_b);
+        if is_valid_solution(&solution) {
+            return solution;
+        } else {
+            // remove order with worst limit price that is selling excess token (to make it less excessive) and try again
+            let excess_token = if context_a.is_excess_before_fees(&context_b) {
+                context_a.address
+            } else {
+                context_b.address
+            };
+            orders.sort_by(price_comparator_for_selling_excess_token(excess_token));
+            orders.pop();
```

Comment:
Could do something like this so it's not sorted every time:
```rust
let comparator = price_comparator_for_selling_excess_token(excess_token);
let index = orders.iter().enumerate().max_by(|left, right| comparator(left.1, right.1).unwrap().0);
orders.swap_remove(index);
```
And we can add a `filter(order_sells_excess_token)` instead of treating orders that don't sell the excess token differently in the the comparator. With the filter we wouldn't `unwrap` in the snippet above and instead break out of the loop early if there is no longer any such order.

Reply:
Very nice suggestion. This made the whole comparator easy enough to inline it.